### DISCUSSION TITLE
Cannot get embedded script working

### DISCUSSION THREAD

**Discussion by gutyex** [ipxe/ipxe#701] [Q&A] [answered]
The goal of my setup is:  
  
Load iPXE via TFTP on regular PXE boot from various machines <- Working
iPXE chainloads a script from an FTP server on the network which I can then modify to change the behaviour without re-compiling  
  
I have successfully modified the configuration from the default pulled from Git to enable FTP, buy my embeded script is not working.  
  
The command used to ccompile is
`make bin/undionly.kpxe EMBED=/home/user/myscript.ipxe`

The contents of myscript.ipxe is
```
#!ipxe

dhcp
chain ftp://nas/boot.ipxe || shell
```
  
I originally tried this without the `|| shell`, then added that to try and get to a shell for troubleshooting. 
All that happens with either variant is that iPXE loads, then the machine continues with the rest of its boot sequence. No errors are shown.

**Comment by stappersg** [comment]
> No errors are shown.

And what can been seen with [pcap](https://en.wikipedia.org/wiki/Pcap)?

**Comment by NiKiZe** [comment, marked as answer]
Have you verified that iPXE starts at all, and that it starts the correct build?

Have you tried running this manually?
Your first likely failure is dhcp, but if so then you should see that fail. (error message) But regardless you should see the attempt.

So the more likely scenario I think is that the script does not actually start with #!ipxe as required, but rather has a BOM char or something else first. Could you run manually and then try to chain into this script? Put it on the FTP.
If the above is the case you will see a link that takes you to https://ipxe.org/2e008000 see the last paragraph on that page.

Another possible but maybe unlikely scenario is that you have disabled scripts in your build. By checking the features shown at startup you could verify this.

To help you further I'd recommend that you upload an image of what you see on screen.

**Reply by gutyex** [reply]
Yes, iPXE is starting and it is starting the version I built - at first I built with the default options from the repository which did not include FTP support, then I re-compiled after enabling FTP in the config and FTP now appears in the features listed when iPXE starts.  

I do not see any indication of network connectivity afte iPXE loads. The Features line appears, and then the machine appears to exit iPXE and return to the rest of it's boot sequence, displaying a 'No bootable devices found' message.

I have not been able to enter a shell where I could attempt to chain the script manually. The script was created with the Ubuntu default text editor application, I guess re-creating it via the terminal with echo "#!ipxe..." > script.ipxe should ensure there are no stray characters?

I will not be back at the place where I am working on this until Thursday, will update with pictures of what I am seeing then.

**Reply by NiKiZe** [reply]
The first step could be to build without the embedded script, which should allow you to press ctrl+b to get a shell.
Here you can run dhcp and the chain command. Again it might be good to chain your script to see what you get. (To verify what/if it is the issue, even if you find a solution)

**Reply by gutyex** [reply]
I think it must be an issue with the encoding of my scripts.  
  
Re-compiling with a fresh pull from Git, only modification being to enable FTP and compiling with `make bin/undionly.kpxe` I was able to get into a shell, acquire a DHCP lease, and fetch the script I was trying to embed via FTP. Running the script caused iPXE to hang.  
Doing the same with the script I was trying to chain from the embedded script gave me `Could not boot: Exec format error`  
  
Re-creating the two scripts directly on the NAS (mounted as an SMB share) with Notepad++ in Windows using UTF-8 encoding and Unix-style newlines corrected those two errors and allowed me to manually load the embedded script which successfully chains the second script. 
  
On Monday I will try re-compiling with the new embedded script file.

**Reply by NiKiZe** [reply]
I think that confirms the encoding issue.

**Reply by gutyex** [reply]
Re-compiling with the new script file embedded is working fine.
